### DISCUSSION TITLE
Is it meant to be possible to debug .NET projects running under .NET Aspire in Docker containers?

### DISCUSSION THREAD

**Discussion by zejji** [microsoft/aspire#6814] [Q&A] [answered]
When working with .NET projects **not** running under .NET Aspire, it is easy to debug .NET code running inside a Docker container using Visual Studio's fast build feature as described here: https://learn.microsoft.com/en-us/visualstudio/containers/edit-and-refresh

I am working on a new distributed application running under .NET Aspire and would like:
- each of the .NET projects to run within their own Docker containers; and
- to be able to set breakpoints in any of the .NET projects and debug my code in Visual Studio.

The reason I need to use Docker containers for my .NET development are:
- I would like to use [Python.NET](https://pythonnet.github.io/) in my .NET services, so would like to install a Python distribution into my .NET Docker containers; and
- I would like to use [Yarp](https://microsoft.github.io/reverse-proxy/) to reverse proxy traffic to other Docker services which are not managed by my Aspire app host, so need the various containers to be on the same Docker network.

However, it is not clear to me from the documentation whether this is intended to be possible and, if so, how. The documentation is currently ambiguous and it would be great if it could be updated if this is indeed possible.

If this is not possible, is the intention that anyone wishing to run .NET projects inside Docker containers use Docker Compose and the standalone Aspire dashboard instead?

**Comment by davidfowl** [comment, marked as answer]
Aspire doesn't run projects in a container today. When we do add this feature, we'll integrate with the IDE for debugging.

**Reply by zejji** [reply]
Thanks for the quick response - looking forward to that when it arrives!

**Reply by runxc1** [reply]
Here it is 18 months later and this is still the most desired missing feature.  The only projects I don't use Aspire with are ones where we have a custom Docker file that installs dependencies needed for the micro service

**Reply by adamint** [reply]
There should be movement on container debug support within the next few aspire versions.

**Reply by DNVTOBS** [reply]
Was looking to see if anyone else wanted this. Good to know I'm not alone haha.

**Comment by michaeljunkin** [comment]
Would love to have that feature, we have some projects that have to run in Linux, would love to be able to debug while using Aspire.

**Comment by ashortland-eqix** [comment]
What's the answer for this?  I've been playing with it and I can run the project in docker using AddDockerfile, but I cannot figure out how to get the debugger attached.  

I've set the stage to base.

Added bind mounts for
/app
/src
/remote_debugger
/.nuget/packages
/.nuget/fallbackpackages
/VSTools
/HotReloadAgent.  

Added labels for 
com.microsoft.created-by
com.microsoft.visual-studio.project-name
com.microsoft.visualstudio.debuggee.arguments
com.microsoft.visualstudio.debuggee.killprogram
com.microsoft.visualstudio.debuggee.program
com.microsoft.visualstudio.debuggee.workingdirectory

Set tty to true

Added environment variables for
DOTNET_USE_POLLING_FILE_WATCHER
ASPNETCORE_LOGGING__CONSOLE__DISABLECOLORS
NUGET_FALLBACK_PACKAGES

Best I can tell this should be everything, but when I set a breakpoint in the project and I have AppHost selected as the startup project I get an alert on my breakpoint that a copy of the file was found in AppHost.dll but it is different so the breakpoint won't be hit.

Any tips on where to go with this would be greatly appreciated, I'm aware this is not a supported scenario at this time, but I have some unique requirements where having this work would be extremely helpful.

**Reply by davidfowl** [reply]
I’m not sure you can do it without IDE help. You have to wait for us to implement this in the ide protocol and in the IDEs

**Reply by ashortland-eqix** [reply]
I was hoping to leverage Aspire for scenarios on Mac where dcproj isn't supported and to streamline picking run modes between host/container for various applications in the same way between full VS and VSCode.

**Reply by davidfowl** [reply]
It’ll be supported, it’s just not supported yet. We have to do the work to make it easy to run the project in a container, and tell the IDE to attach to the remote debugger running in the container.

@danegsta and @dbreshears can comment on this but we just haven’t prioritized this yet

**Reply by ashortland-eqix** [reply]
@danegsta @dbreshears If either of you can provide any insight into how this could work/if it's even possible without IDE updates that would be great.  I'd love to get this working as it would be a huge help with some of my workflows.

**Reply by karolz-ms** [reply]
@ashortland-eqix running and debugging projects in containers is not possible without new functionality in the IDE and Aspire app host libraries. There is no workaround today, sorry.

Could you share what is for you the benefit of debugging of a project running inside a container? Especially as opposed to debugging the app running (the whole app) on Linux/WSL?

**Reply by davidfowl** [reply]
We also don't support running in WSL via aspire 😄. That would be a huge win as well.

**Reply by ashortland-eqix** [reply]
@karolz-ms 
In a nutshell, the scenario it would help me with is a consistent/fast development experience.

When you have devs using VS, others using VS Code, some using Windows, and others using Mac with a basic project all is well.  Once you get into more complex projects that require Redis, SQL Server/Postgres, a separate project to manage/apply db migrations, and need to orchestrate that flow things get much more difficult and workflows start fragmenting pretty quickly.  

Aspire is amazing and helps with almost all of this, but there are still scenarios where some apps in these flows are only capable of being run on Windows, and others that are only going to run on Linux containers in a real deployment.  The team has done an excellent job at streamlining API's and unifying .NET API's, but there are nuanced differences that you don't see until running on the actual target platform.

If Aspire could allow the developer to quickly toggle between running various projects in the solution on the host, in a container, etc. it would provide a truly universal solution for orchestrating these more complex scenarios regardless of the developer's choice of IDE and platform.

**Reply by ArnoldIsHereToday** [reply]
I am migrating .net  4.8 framework console applications to modern .Net containers that are cross platform. So I want to know if projects run correctly on Linux.
Whenever possible you want to mimick production as much as possible (and practical) during development. 

When I am targeting Kubernetes: and it would be easy to use and quick to start; I would like aspire to debug projects inside Kubernetes pods. 

So using docker (compose) would get closer to mimicking production.

**Reply by karolz-ms** [reply]
@ashortland-eqix @ArnoldIsHereToday appreciate you both take time to explain why running projects inside containers would help you. Makes a lot of sense, thank you.

Let me share some thoughts on what you can do today. This is my personal opinion, not Aspire team "official stance" or anything like that. Hope it helps!

When porting things from "Windows server environment" to "containers in a Kubernetes cluster", problems can be roughly classified into two kinds. The first kind stems from issues with how the container/pod, and the process inside it, are configured. Here I mean both problems with how container image is built, as well as what data is injected into it, and how. Call it "setup" problems. You tend to have them more during initial phase of development; once you get your build/CI pipelines going, they tend to be relatively rare. You can catch and diagnose some of these problems by running projects inside a container during development, using Docker/Podman/Docker Compose. But you won't be able to catch all setup problems like that. You need to have integration tests that deploy things to real Kubernetes clusters, and good, verbose enough logging to diagnose problems when these tests fail. Which you need for production too, because you won't be debugging things in production. For all these reasons running inside containers during development helps, but is not a complete solution. Again, you want integration tests.

The second kind of problems (the majority in my experience) is "this used to work on Windows, but does not work on Linux" kind of problem. To catch and debug those problems it is enough to run the application on Linux/WSL. You can do this TODAY with Aspire and VS Code. The experience is admittedly a bit behind VS, but we are committed to making it better--if you encounter problems, we'd love to hear about them, please open issues. I also agree with @davidfowl that running projects inside WSL would be really helpful here. No promises for now 😉 

The other thing that you can do today is to leverage the fact that app host is using C# code to configure the application. It is easy to write some code and use configuration/environment variables to put the application together in a different way, or use subset of projects and supporting resources for particular run. So e.g. if running on non-Windows OS, you can exclude projects that are know to require Windows etc.

**Reply by ArnoldIsHereToday** [reply]
Thank you very much for your response. I am still in the early stages of using aspire with visual studio. Utilizing aspire to make integration test is something I did consider, but have not yet had time for.  I do love the "start from scratch" experience you get with infra as C# code and options to include / exclude components via code.

From what you are saying: I can switch from Visual Studio to VS code now and already run everything on WSL. Not as a container, but on Linux, which is already a lot better for my scenario.. 

The remark about adding  "running projects inside WSL would be really helpful here" only applies to Visual Studio then, since VS code already supports it.

 I will try that out.

**Reply by ashortland-eqix** [reply]
@karolz-ms Thanks for the detailed response to this.  I'm happy to do some work on this, but need a point in the right direction.

In my previous comment on this thread I've already started looking at how this works with running from dcproj and noted that there's a generated compose file that swaps the entrypoint and slips several volumes and labels into the container.  It wasn't too difficult to do the same with Aspire launching the application within a container so I believe I have all the requirements in place from the container perspective. 

This is where I'm stuck.

I was watching docker events running a project with dcproj and see that there are some execs running against the container once it's spun up.  My assumption is this is VS attaching to the debugger and using the labels to actually launch the application within the container.  It doesn't seem like it would be difficult to use DCP to do the same, but I haven't been able to find any clear information about exactly what's happening at this point without requiring the user to manually attach to the process in the container.

I see the DistrolessHelper being invoked with a server argument and an ID, my guess is the IDE generates the ID and invoked the DistrolessHelper with it to orchestrate a pipe of some kind that the debugger will connect to, but that's largely a guess.

**Reply by karolz-ms** [reply]
@ashortland-eqix I will let my coworker, @NCarlsonMSFT, comment on the feasibility of leveraging VS dcproj support with Aspire. As far as I know it *might* work with some extra configuration, but we have not made any effort to ensure that it does, so you are in uncharted territories as of today.

**Reply by NCarlsonMSFT** [reply]
@ashortland-eqix the labels you see are part of the configuration story for the .dcproj ([docs](https://learn.microsoft.com/en-us/visualstudio/containers/docker-compose-properties?view=vs-2022#docker-compose-file-labels)). Adding them to an arbitrary container won't get you debugging.

**Note this is all spit balling and mainly left as an exercise for the reader:**
I don't know enough about extensibility on the aspire side, but if you're watching docker events you may already be able to get the information you need to create a more scripted approach to attach to your aspire containers (i know it's not as good as F5, but it's better than manually attaching each time).

The steps for creating an attach to container configuration file are [here](https://github.com/Microsoft/MIEngine/wiki/Offroad-Debugging-of-.NET-Core-on-Linux---OSX-from-Visual-Studio#attaching) and the steps for invoking it are [here](https://github.com/Microsoft/MIEngine/wiki/Offroad-Debugging-of-.NET-Core-on-Linux---OSX-from-Visual-Studio#start-debugging). Assuming you are able to identify your containers and the process Ids you're running in them you should be able to generate a config file per process, and invoke the command for each of those [command automation docs](https://learn.microsoft.com/en-us/dotnet/api/envdte._dte.executecommand?view=visualstudiosdk-2019)

**Reply by NCarlsonMSFT** [reply]
@ashortland-eqix, @ArnoldIsHereToday You've inspired me and I was able to get a POC working locally. Sadly, I can't just publish it as an example b/c i grabbed some existing parts i had lying around that i don't know the licensing of :/ Here are the parts that i know i can give away:

A) Utilities to get a cross-process COM object for automating the VS with your project:
```C#
private static EnvDTE.DTE GetDte(string projectName)
{
    var procs = GetVSProcessIds();

    foreach (var pid in procs)
    {
        var dte = GetDte(pid);
        if (dte.Solution == null || dte.Solution.Projects == null || dte.Solution.Projects.Count <= 0)
        {
            continue;
        }

        if (dte.Solution.Projects.Cast<EnvDTE.Project>().Any(p => p.Name == projectName))
        {
            return dte;
        }
    }

    return null;
}

private static int[] GetVSProcessIds()
{
    List<int> processIds = new List<int>();
    foreach (var process in System.Diagnostics.Process.GetProcessesByName("devenv"))
    {
        processIds.Add(process.Id);
    }
    return processIds.ToArray();
}

private static EnvDTE.DTE GetDte(int pid)
{
    IntPtr numFetched = IntPtr.Zero;
    IRunningObjectTable runningObjectTable;
    IEnumMoniker monikerEnumerator;
    IMoniker[] monikers = new IMoniker[1];

    //get Running object table, make a pinvoke call
    NativeMethods.GetRunningObjectTable(0, out runningObjectTable);

    //get an enumerator for entries
    runningObjectTable.EnumRunning(out monikerEnumerator);
    monikerEnumerator.Reset();

    string runningObjectNameExpected = "!VisualStudio.DTE.17.0:" + pid;

    while (monikerEnumerator.Next(1, monikers, numFetched) == 0)
    {
        IBindCtx ctx;
        NativeMethods.CreateBindCtx(0, out ctx);

        string runningObjectName;
        monikers[0].GetDisplayName(ctx, null, out runningObjectName);

        if (runningObjectName.Equals(runningObjectNameExpected, StringComparison.OrdinalIgnoreCase))
        {
            object runningObjectVal;
            runningObjectTable.GetObject(monikers[0], out runningObjectVal);
            return (EnvDTE.DTE)runningObjectVal;
        }
    }

    return null;
}
```
B) Find the running container:
```C#
private static string GetContainerId(string imageName)
{
    ProcessStartInfo processStartInfo = new ProcessStartInfo
    {
        FileName = "docker",
        Arguments = $"ps -q --filter \"ancestor={imageName}\"",
        RedirectStandardOutput = true,
        UseShellExecute = false,
        CreateNoWindow = true
    };

    while (true)
    {
        Process docker = Process.Start(processStartInfo);
        docker.WaitForExit();
        string[] outputLines = docker.StandardOutput.ReadToEnd()?.Split(new[] { Environment.NewLine }, StringSplitOptions.RemoveEmptyEntries) ?? Array.Empty<string>();

        if (docker.ExitCode != 0 || outputLines.Length == 0)
        {
            System.Threading.Thread.Sleep(1000);
            continue;
        }

        return outputLines.First().Trim();
    }
}
```
C) Invoke debugging in the container:
```C#
string containerId = GetContainerId(imageName);
using (new MessageFilter())
{
    var dte = GetDte(projectName);

    string config = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString("N") + ".json");

    // See https://github.com/Microsoft/MIEngine/wiki/Offroad-Debugging-of-.NET-Core-on-Linux---OSX-from-Visual-Studio#attaching
    File.WriteAllText(config, $$"""
         {
          "version": "0.2.0",
          "adapter": "docker.exe",
          "adapterArgs": "exec -i {{containerId}} /remote_debugger/vsdbg --interpreter=vscode",
          "configurations": [
            {
              "name": ".NET Core Docker Attach",
              "type": "coreclr",
              "request": "attach",
              // Just assuming the target process is the entry point and has pid 1
              "processId": 1
            }
          ]
        }
        """);

    dte.ExecuteCommand("DebugAdapterHost.Launch", $"/LaunchJson:\"{config}\" /EngineGuid:541B8A8A-6081-4506-9F0A-1CE771DEBC04");

}
```

To make this work I added a .NET 4.7 console app to my solution (and annotated it with  STAThreadAttribute for COM) and made that an additional start-up project. When debugging started my bonus console app would wait for the container using my known base image and then find the correct VS and manually invoke the attach command for the container. The biggest exercise I've left to the reader is implementing the IMessageFilter that makes cross process COM more reliable, but a quick [bing search](https://www.bing.com/search?q=00000016-0000-0000-C000-000000000046) found several other people with examples for that,

Also note that this assumes vsdbg is available in the container at `/remote_debugger/vsdbg`, i did it as a volume mount but you could also have a special stage for debugging that adds it to the image.

**Reply by ashortland-eqix** [reply]
This is amazing!  Thank you, I'll definitely be digging deeper into this!

**Reply by NCarlsonMSFT** [reply]
@ashortland-eqix, @ArnoldIsHereToday if you're still interested, I was a bit stuck on this, and having spent far too much time on it over the weekend, I now do have a sample repo you can look at: https://github.com/NCarlsonMSFT/ContainerDebuggingInAspire

The repo is taking the basic Aspire start app, throwing everything into an SDK container, and then adding auto-attaching and HTTPS.

It has some rough edges but should give you an idea of what is possible.

**Reply by CZEMacLeod** [reply]
@NCarlsonMSFT I just did almost the same thing for debugging IIS Express projects using [MSBuild.SDK.SystemWeb](https://github.com/CZEMacLeod/MSBuild.SDK.SystemWeb). I put together some stuff to generically attach the VS debugger to resources automatically and have it working with IIS Express, and node available at [C3D.Extensions.Aspire.IISExpress](https://github.com/CZEMacLeod/C3D.Extensions.Aspire.IISExpress).
I haven't published this as packages yet (need to renew a signing certificate) - but will do so at some point soon.
